src/middlewared/middlewared/plugins/jbof/redfish/client.py

```diff
@@ -341,3 +346,240 @@ def link_status(self, uri):
+            else:
+                with Client(f'ws+unix://{MIDDLEWARE_RUN_DIR}/middlewared-internal.sock', py_exceptions=True) as c:
+                    jbofs = c.call('jbof.query', filters)
+
+            for jbof in filter_list(jbofs, filters, options):
+                base_urls = []
+                for key in ['mgmt_ip1', 'mgmt_ip2']:
+                    if mgmt_ip := jbof.get(key):
+                        base_urls.append(f'https://{mgmt_ip}')
+                redfish = await cls.create(base_urls, jbof['mgmt_username'], jbof['mgmt_password'])
+                cls.cache_set(uuid, redfish)
+
+            return redfish
+
+    async def _login(self, base_url, username=None, password=None, authtype=None):
+        """Login using the specified path and credentials."""
+        self.username = username if username else self.username
+        self.password = password if password else self.password
+        if authtype:
+            self.authtype = AuthMethod.authtype_to_enum(authtype)
+
+        async with aiohttp.ClientSession(base_url, timeout=self.timeout) as session:
+            if self.authtype == AuthMethod.BASIC:
+                auth = aiohttp.BasicAuth(self.username, self.password)
+                async with session.get(self.login_url, ssl=False, auth=auth) as response:
+                    if not response.ok:
+                        raise InvalidCredentialsError('Could not authenticate credentials supplied')
+                newsession = aiohttp.ClientSession(base_url, timeout=self.timeout, raise_for_status=True, auth=auth)
+            elif self.authtype == AuthMethod.SESSION:
+                data = {'UserName': self.username, 'Password': self.password}
+                async with session.post(self.login_url, ssl=False, json=data) as response:
+                    if not response.ok:
+                        raise InvalidCredentialsError('Could not authenticate credentials supplied')
+                    auth_token = response.headers.get('X-Auth-Token')
+                    if auth_token:
+                        # Save the Location for logout purposes
+                        self.session_location[base_url] = response.headers.get('Location')
+                newsession = aiohttp.ClientSession(base_url, timeout=self.timeout, raise_for_status=True, headers={'X-Auth-Token': auth_token})
+            else:
+                raise ValueError('Invalid auth supplied:', authtype)
+        # Save the session for reuse
+        self._add_session(base_url, newsession)
+        return newsession
+
+    async def _logout(self, base_url, session):
+        if self.authtype == AuthMethod.BASIC:
+            self.auth = None
+        elif self.authtype == AuthMethod.SESSION:
+            if location := self.session_location.get(base_url):
+                async with session.delete(location, ssl=False) as response:
+                    if response.ok:
+                        del self.session_location[base_url]
```

Comment:
Is it possible to fail with KeyError here?

Reply:
I don't believe so, as we just did the dict.get, but can protect anyway.